### DISCUSSION TITLE
Vue

### DISCUSSION THREAD

**Discussion by wobsoriano** [css-hooks/css-hooks#37] [Ideas]
cuz why not?

```vue
<script setup lang="ts">
import { css, hooks } from './css'
</script>

<template>
  <component is="style" v-html="hooks"></component>
  <button
    :style="css({
      border: 0,
      margin: 0,
      padding: '0.75rem 1rem',
      fontFamily: 'sans-serif',
      fontSize: '1rem',
      background: '#333',
      color: '#fff',
      borderRadius: '0.5rem',
      '&:hover': { background: '#444' },
    })"
  >
    CSS Hooks
  </button>
</template>
```

with Nuxt

```vue
<script setup lang="ts">
import { css, hooks } from './css'
</script>

<template>
  <Head>
    <Style type="text/css" :children="hooks"></Style>
  </Head>
</template>
```

**Comment by wobsoriano** [comment]
Would it be worth it to add a Vue package? Code is simple

```ts
import { buildHooksSystem, genericStringify } from '@css-hooks/core'
import { CSSProperties } from 'vue'


export const createHooks = buildHooksSystem<CSSProperties>(genericStringify)
```

**Comment by nsaunders** [comment]
Cool! I'm not deeply familiar with Vue, but if CSS Hooks can provide some benefit there, I also don't see why not.

It's great to see that you've already reviewed the core library and that a new framework integration might be just that easy.

I do have some questions:
1. Does Vue turn style objects into strings at all, or is that one additional step that the `css` function will need to include for Vue?
2. If it _does_ convert the style object, do you know where to find that logic in Vue's codebase? We'd want to ensure that it lines up with `genericStringify`, or else create a new stringify function that aligns.
3. Would you mind putting together a demo on CodeSandbox (or equivalent) please? For now, it would just use the core library directly. Then, once we are satisfied with it, we can publish the `createHooks` function in a new `@css-hooks/vue` package.

Thanks for your interest!

**Reply by wobsoriano** [reply]
> Does Vue turn style objects into strings at all, or is that one additional step that the CSS function will need to include for Vue?

`style` properties in Vue templates accept strings, objects, or an array of objects. So I think `genericStringify` should do the job? or even without it? Any of the following are valid in Vue:

1. `<div style="text-align: center; color: red;">hello</div>`
2. `<div :style="{ textAlign: 'center', color: 'red' }">hello</div>`
3. `<div :style="[{ textAlign: 'center' }, { color: 'red' }]">hello</div>`

Since the style prop accepts an array, we can probs update the type to 

```ts
import { CSSProperties } from 'vue'

const createHooks = buildHooksSystem<CSSProperties | Array<CSSProperties>>(genericStringify)
```

So users can still do this with good typing support

```vue
<template>
  <div :style="[css({ color: 'white'}), css({ backgroundColor: 'black' })]">
    some text
  </div>
</template>
```

> If it does convert the style object, do you know where to find that logic in Vue's codebase? We'd want to ensure that it lines up with genericStringify, or else create a new stringify function that aligns.

I am not sure where exactly is that function, but it's most likely in the ff. lines:

- https://github.com/vuejs/core/blob/a6503e3e77b04a1ab8c07078e07091b06d51131d/packages/runtime-dom/src/modules/style.ts#L7
- https://github.com/vuejs/core/blob/a6503e3e77b04a1ab8c07078e07091b06d51131d/packages/shared/src/normalizeProp.ts#L5

> Would you mind putting together a demo on CodeSandbox (or equivalent) please? For now, it would just use the core library directly. Then, once we are satisfied with it, we can publish the createHooks function in a new @css-hooks/vue package.

Sure! Here's the [link](https://codesandbox.io/p/devbox/css-hooks-vue-cc8yfq) to the demo. 

If you think everything looks good, I can open a PR!

**Comment by nsaunders** [comment]
Regarding the "CSS properties" type, that would need to be `CSSProperties` only (no `Array<CSSProperties>`). That's because, when we introduce nested style objects, there's no longer a single merge strategy that makes sense for every use case.

I also notice in that demo that there are a few warnings about deprecated "@vnode-* hooks"? Do you understand that issue, and is there anything we can do?

Other than that, I don't have much feedback at the moment, except to say this looks very cool! Thank you for reaching out and putting this demo together... I really had no idea this project would be relevant for Vue users.

I think we are just about ready take the first step of adding a Vue [example](https://github.com/css-hooks/css-hooks/tree/master/examples). There are still other things to figure out, like documentation (since we're focused on React/JSX up to this point, and the docs are structured accordingly), but we'll just take it one step at a time.

Thanks again @wobsoriano!

**Reply by wobsoriano** [reply]
> I also notice in that demo that there are a few warnings about deprecated "@vnode-* hooks"? Do you understand that issue, and is there anything we can do?

I think that one's caused by [CodeSandbox vue devtools](@codesandbox/vue-preview). It works on my machine 😆 

For docs, I think it's gonna be the same as the other examples except no. 3 (Add static CSS). I'm thinking of something like:

## 3. Add static CSS

The hooks export from the module created in Step 2 consists of a static CSS string that must be included somewhere on the page. This could be achieved in any number of ways, but the easiest is simply to use Vue's built-in [`component`](https://vuejs.org/guide/essentials/component-basics.html#dynamic-components) element, e.g.:

```vue
<script setup lang="ts">
import  { hooks } from './css'
</script>

<template>
  <component is="style" v-html="hooks"></component>
</template>
```

For Nuxt users, use the built-in [Style](https://nuxt.com/docs/getting-started/seo-meta#components) component, e.g.:

```vue
<script setup lang="ts">
import { hooks } from './css'
</script>

<template>
  <Head>
    <Style type="text/css" :children="hooks"></Style>
  </Head>
</template>
```

**Reply by gemmadlou** [reply]
Nice @wobsoriano. Thank you putting this together. I want to have a try of it too. The sandbox has a 404 but your code examples are simple enough to follow.

**Reply by wobsoriano** [reply]
@gemmadlou that's weird. You can try this demo from [stackblitz](https://stackblitz.com/edit/vitejs-vite-281byb?file=src%2FApp.vue) instead.

**Reply by gemmadlou** [reply]
Sorry @wobsoriano. Just saw stack blitz. The syntax looks way easier and explicit now. 🙏

**Comment by nsaunders** [comment]
Hi @wobsoriano, sorry for the delay! Thanks again for exploring this integration. I'd be happy to entertain a PR whenever you have time to put it together.

**Comment by nsaunders** [comment]
I'm going to close this for now as it appears to be stale. Please let me know if/when there is renewed interest in this.
